# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_ids, target_ids = [], []
    source_split = source_text.split("\n")
    for s in source_split:
        ids = [source_vocab_to_int[word] for word in s.split()]
        source_ids.append(ids)
    target_split = target_text.split("\n")
    for s in target_split:
        ids = [target_vocab_to_int[word] for word in s.split()]
        ids.append(target_vocab_to_int['<EOS>'])
        target_ids.append(ids)

    return (source_ids, target_ids)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    _input = tf.placeholder(tf.int32, [None, None], name="input")
    _targets = tf.placeholder(tf.int32, [None, None], name="targets")
    _lr = tf.placeholder(tf.float32, name="learning_rate")
    _keep_prob = tf.placeholder(tf.float32, name="keep_prob")
    _target_sequence_length = tf.placeholder(tf.int32, (None,), name="target_sequence_length")
    _max_target_sequence_length = tf.reduce_max(_target_sequence_length, name="max_target_len")
    _source_sequence_length = tf.placeholder(tf.int32, (None,), name="source_sequence_length")
    return (
        _input, _targets, _lr, _keep_prob, _target_sequence_length, 
        _max_target_sequence_length, _source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    # worth mentioning that I had a hard time understanding strided_slice, but basically
    # you are drawing a rectangle with two coordinates, [0,0] is the top left corner
    # [batch_size, -1] is the bottom right corner, and [1,1] just means your stride
    # so (in this case, keep everything)
    ending = tf.strided_slice(target_data, [0,0], [batch_size, -1], [1,1])
    # this basically creates a rank 2 tensor of '<GO'>, that is batch_size x 1, like a vertical vector
    # it then pre-pends it to the ending tensor which is batch_size x (len(target[1]) - 1)
    # and it does this along the column axis (axis = 1)
    decoder_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return decoder_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    # Taking the input, you use the vocabulary size, and encoding_embedding size as two of
    # the i don't know 3?? dimensions, this needs futher study
    enc_embedding = tf.contrib.layers.embed_sequence(
        rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    # encoder
    def make_cell(rnn_size):
        # construct our cell and initialize
        # made of LSTM cells
        enc_cell = tf.contrib.rnn.DropoutWrapper(
            tf.contrib.rnn.LSTMCell(
                rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2)),
            input_keep_prob=keep_prob)
        
        return enc_cell # this is just a single layer of our encoder
    
    # now to make the full multi_rnn_cell of num_layer encoding cells
    enc_cell = tf.contrib.rnn.MultiRNNCell(
        [make_cell(rnn_size) for _ in range(num_layers)])
    # print(enc_cell)
    
    # lets embed the input
    enc_output, enc_state = tf.nn.dynamic_rnn(
        enc_cell, enc_embedding, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return (enc_output, enc_state)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    # helper for the training process.  used by BasicDecoder to read inputs.
    # whatever this means :P
    # print(max_summary_length.shape, max_summary_length.dtype)
    training_helper = tf.contrib.seq2seq.TrainingHelper(
        inputs=dec_embed_input, sequence_length=target_sequence_length, time_major=False)
    # basic decoder
    # ?? get the max target sequence length
    #max_target_sequence_length = tf.reduce_max(target_sequence_length)
    # wrapping a dropout for keep probability
    # print(dec_cell)
    dec_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, input_keep_prob=keep_prob)
    # print(dec_cell)
    # make the training decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(
        dec_cell, training_helper, encoder_state, output_layer)
    # perform dynamic decoding using the decoder
    x = tf.contrib.seq2seq.dynamic_decode(
        training_decoder, impute_finished=True, maximum_iterations=max_summary_length)
    basic_decoder_output = x[0]

    return basic_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    # create a 1d tensor of <GO> codes as our start tokens
    start_tokens = tf.tile(
        tf.constant([start_of_sequence_id], dtype=tf.int32),
        [batch_size], name="start_tokens")
    
    # Helper for the inference process
    # This is an interesting function, it needs a vector of start_sequences, and a scalar
    # of the end_of_sequence
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(
        dec_embeddings, start_tokens, end_of_sequence_id)
    

    # adding dropout wrapper to the decode cell
    dec_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, input_keep_prob=keep_prob)
    
    # Basic decoder    
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(
        dec_cell, inference_helper, encoder_state, output_layer)
    
    # Perform dynamic decoding using the decoder
    x = tf.contrib.seq2seq.dynamic_decode(
        inference_decoder, impute_finished=True, maximum_iterations=max_target_sequence_length)
    
    basic_decoder_output = x[0]
    
    return basic_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # 1. Decoder embedding
    dec_embeddings = tf.Variable(tf.random_uniform(
        [target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        # construct our cell and initialize
        # made of LSTM cells
        dec_cell = tf.contrib.rnn.DropoutWrapper(
            tf.contrib.rnn.LSTMCell(
                rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2)),
            input_keep_prob=keep_prob)
        
        return dec_cell # this is just a single layer of our encoder
    
    # Stack layers
    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])  
    
    # 3. Dense layer to translate decoders outputs at each time step into a choice from the
    # target vocabulary
    output_layer = Dense(
        target_vocab_size, kernel_initializer=tf.truncated_normal_initializer(mean=0.0, stddev=0.1))
    
    # 4. Set up training decoder
    with tf.variable_scope("decode"):
        train = decoding_layer_train(
            encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length,
            output_layer, keep_prob)
    
    # 5. Set up inference decoder
    with tf.variable_scope("decode", reuse=True):
        start_of_sequence_id = target_vocab_to_int['<GO>']
        end_of_sequence_id = target_vocab_to_int['<EOS>']
        infer = decoding_layer_infer(
            encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
            max_target_sequence_length, target_vocab_size, output_layer, batch_size, keep_prob)
    
    return (train, infer)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # 1. Encode the input
    _, enc_state = encoding_layer(
        input_data, rnn_size, num_layers, keep_prob, source_sequence_length, source_vocab_size,
        enc_embedding_size)
    # 2. Process target data
    dec_input = process_decoder_input(
        target_data, target_vocab_to_int, batch_size)
    
    # 3. Decode the encoded input using the decoding layer
    train, infer = decoding_layer(
        dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size,
        num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)
    
    return train, infer


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [22]:
# Number of Epochs
epochs = 20
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 150
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.75
display_step = 50

### Build the Graph
Build the graph using the neural network you implemented.

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [24]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [25]:
from tqdm import tqdm
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in tqdm(range(epochs)):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

  0%|          | 0/20 [00:00<?, ?it/s]

Epoch   0 Batch   50/538 - Train Accuracy: 0.3299, Validation Accuracy: 0.3906, Loss: 2.9994
Epoch   0 Batch  100/538 - Train Accuracy: 0.4225, Validation Accuracy: 0.4751, Loss: 2.5386
Epoch   0 Batch  150/538 - Train Accuracy: 0.4643, Validation Accuracy: 0.5051, Loss: 2.2838
Epoch   0 Batch  200/538 - Train Accuracy: 0.4641, Validation Accuracy: 0.5140, Loss: 2.0890
Epoch   0 Batch  250/538 - Train Accuracy: 0.4637, Validation Accuracy: 0.5238, Loss: 1.9422
Epoch   0 Batch  300/538 - Train Accuracy: 0.4712, Validation Accuracy: 0.4877, Loss: 1.6597
Epoch   0 Batch  350/538 - Train Accuracy: 0.4950, Validation Accuracy: 0.5137, Loss: 1.5511
Epoch   0 Batch  400/538 - Train Accuracy: 0.5086, Validation Accuracy: 0.5302, Loss: 1.3874
Epoch   0 Batch  450/538 - Train Accuracy: 0.5372, Validation Accuracy: 0.5421, Loss: 1.2669
Epoch   0 Batch  500/538 - Train Accuracy: 0.5259, Validation Accuracy: 0.5336, Loss: 1.1211


  5%|▌         | 1/20 [00:55<17:39, 55.76s/it]

Epoch   1 Batch   50/538 - Train Accuracy: 0.5236, Validation Accuracy: 0.5513, Loss: 1.0481
Epoch   1 Batch  100/538 - Train Accuracy: 0.5350, Validation Accuracy: 0.5616, Loss: 0.9892
Epoch   1 Batch  150/538 - Train Accuracy: 0.5516, Validation Accuracy: 0.5497, Loss: 0.9419
Epoch   1 Batch  200/538 - Train Accuracy: 0.5574, Validation Accuracy: 0.5614, Loss: 0.8972
Epoch   1 Batch  250/538 - Train Accuracy: 0.5514, Validation Accuracy: 0.5723, Loss: 0.8655
Epoch   1 Batch  300/538 - Train Accuracy: 0.5844, Validation Accuracy: 0.5760, Loss: 0.8136
Epoch   1 Batch  350/538 - Train Accuracy: 0.5755, Validation Accuracy: 0.5653, Loss: 0.8178
Epoch   1 Batch  400/538 - Train Accuracy: 0.5701, Validation Accuracy: 0.5877, Loss: 0.7838
Epoch   1 Batch  450/538 - Train Accuracy: 0.6068, Validation Accuracy: 0.6014, Loss: 0.7841
Epoch   1 Batch  500/538 - Train Accuracy: 0.6087, Validation Accuracy: 0.6065, Loss: 0.6852


 10%|█         | 2/20 [01:51<16:42, 55.68s/it]

Epoch   2 Batch   50/538 - Train Accuracy: 0.5865, Validation Accuracy: 0.6087, Loss: 0.7234
Epoch   2 Batch  100/538 - Train Accuracy: 0.5805, Validation Accuracy: 0.6069, Loss: 0.7067
Epoch   2 Batch  150/538 - Train Accuracy: 0.6027, Validation Accuracy: 0.6113, Loss: 0.6879
Epoch   2 Batch  200/538 - Train Accuracy: 0.6100, Validation Accuracy: 0.6135, Loss: 0.6682
Epoch   2 Batch  250/538 - Train Accuracy: 0.5908, Validation Accuracy: 0.6099, Loss: 0.6586
Epoch   2 Batch  300/538 - Train Accuracy: 0.6112, Validation Accuracy: 0.6064, Loss: 0.6223
Epoch   2 Batch  350/538 - Train Accuracy: 0.6064, Validation Accuracy: 0.6200, Loss: 0.6349
Epoch   2 Batch  400/538 - Train Accuracy: 0.6114, Validation Accuracy: 0.6165, Loss: 0.6082
Epoch   2 Batch  450/538 - Train Accuracy: 0.6183, Validation Accuracy: 0.6158, Loss: 0.6211
Epoch   2 Batch  500/538 - Train Accuracy: 0.6399, Validation Accuracy: 0.6303, Loss: 0.5470


 15%|█▌        | 3/20 [02:46<15:45, 55.63s/it]

Epoch   3 Batch   50/538 - Train Accuracy: 0.6326, Validation Accuracy: 0.6374, Loss: 0.5921
Epoch   3 Batch  100/538 - Train Accuracy: 0.6236, Validation Accuracy: 0.6413, Loss: 0.5747
Epoch   3 Batch  150/538 - Train Accuracy: 0.6535, Validation Accuracy: 0.6355, Loss: 0.5664
Epoch   3 Batch  200/538 - Train Accuracy: 0.6408, Validation Accuracy: 0.6381, Loss: 0.5531
Epoch   3 Batch  250/538 - Train Accuracy: 0.6113, Validation Accuracy: 0.6394, Loss: 0.5652
Epoch   3 Batch  300/538 - Train Accuracy: 0.6499, Validation Accuracy: 0.6474, Loss: 0.5428
Epoch   3 Batch  350/538 - Train Accuracy: 0.6542, Validation Accuracy: 0.6436, Loss: 0.5404
Epoch   3 Batch  400/538 - Train Accuracy: 0.6466, Validation Accuracy: 0.6518, Loss: 0.5287
Epoch   3 Batch  450/538 - Train Accuracy: 0.6438, Validation Accuracy: 0.6641, Loss: 0.5340
Epoch   3 Batch  500/538 - Train Accuracy: 0.6683, Validation Accuracy: 0.6596, Loss: 0.4745


 20%|██        | 4/20 [03:42<14:49, 55.59s/it]

Epoch   4 Batch   50/538 - Train Accuracy: 0.6646, Validation Accuracy: 0.6664, Loss: 0.5057
Epoch   4 Batch  100/538 - Train Accuracy: 0.6484, Validation Accuracy: 0.6726, Loss: 0.4928
Epoch   4 Batch  150/538 - Train Accuracy: 0.6777, Validation Accuracy: 0.6713, Loss: 0.4815
Epoch   4 Batch  200/538 - Train Accuracy: 0.6691, Validation Accuracy: 0.6795, Loss: 0.4647
Epoch   4 Batch  250/538 - Train Accuracy: 0.6469, Validation Accuracy: 0.6774, Loss: 0.4749
Epoch   4 Batch  300/538 - Train Accuracy: 0.6749, Validation Accuracy: 0.6784, Loss: 0.4540
Epoch   4 Batch  350/538 - Train Accuracy: 0.6819, Validation Accuracy: 0.6815, Loss: 0.4478
Epoch   4 Batch  400/538 - Train Accuracy: 0.6726, Validation Accuracy: 0.6779, Loss: 0.4394
Epoch   4 Batch  450/538 - Train Accuracy: 0.6842, Validation Accuracy: 0.6887, Loss: 0.4439
Epoch   4 Batch  500/538 - Train Accuracy: 0.6983, Validation Accuracy: 0.7005, Loss: 0.3926


 25%|██▌       | 5/20 [04:38<13:54, 55.64s/it]

Epoch   5 Batch   50/538 - Train Accuracy: 0.6883, Validation Accuracy: 0.6994, Loss: 0.4193
Epoch   5 Batch  100/538 - Train Accuracy: 0.6736, Validation Accuracy: 0.7072, Loss: 0.4052
Epoch   5 Batch  150/538 - Train Accuracy: 0.7055, Validation Accuracy: 0.7024, Loss: 0.4040
Epoch   5 Batch  200/538 - Train Accuracy: 0.7021, Validation Accuracy: 0.7077, Loss: 0.3926
Epoch   5 Batch  250/538 - Train Accuracy: 0.6871, Validation Accuracy: 0.7072, Loss: 0.4013
Epoch   5 Batch  300/538 - Train Accuracy: 0.6938, Validation Accuracy: 0.7086, Loss: 0.3908
Epoch   5 Batch  350/538 - Train Accuracy: 0.6959, Validation Accuracy: 0.7072, Loss: 0.3833
Epoch   5 Batch  400/538 - Train Accuracy: 0.7072, Validation Accuracy: 0.7083, Loss: 0.3810
Epoch   5 Batch  450/538 - Train Accuracy: 0.7245, Validation Accuracy: 0.7202, Loss: 0.3813
Epoch   5 Batch  500/538 - Train Accuracy: 0.7209, Validation Accuracy: 0.7273, Loss: 0.3425


 30%|███       | 6/20 [05:33<12:59, 55.71s/it]

Epoch   6 Batch   50/538 - Train Accuracy: 0.7121, Validation Accuracy: 0.7235, Loss: 0.3716
Epoch   6 Batch  100/538 - Train Accuracy: 0.7195, Validation Accuracy: 0.7356, Loss: 0.3577
Epoch   6 Batch  150/538 - Train Accuracy: 0.7229, Validation Accuracy: 0.7200, Loss: 0.3612
Epoch   6 Batch  200/538 - Train Accuracy: 0.7357, Validation Accuracy: 0.7367, Loss: 0.3492
Epoch   6 Batch  250/538 - Train Accuracy: 0.7150, Validation Accuracy: 0.7376, Loss: 0.3560
Epoch   6 Batch  300/538 - Train Accuracy: 0.7230, Validation Accuracy: 0.7454, Loss: 0.3466
Epoch   6 Batch  350/538 - Train Accuracy: 0.7333, Validation Accuracy: 0.7425, Loss: 0.3468
Epoch   6 Batch  400/538 - Train Accuracy: 0.7435, Validation Accuracy: 0.7449, Loss: 0.3402
Epoch   6 Batch  450/538 - Train Accuracy: 0.7547, Validation Accuracy: 0.7583, Loss: 0.3439
Epoch   6 Batch  500/538 - Train Accuracy: 0.7521, Validation Accuracy: 0.7567, Loss: 0.3008


 35%|███▌      | 7/20 [06:29<12:05, 55.81s/it]

Epoch   7 Batch   50/538 - Train Accuracy: 0.7389, Validation Accuracy: 0.7633, Loss: 0.3268
Epoch   7 Batch  100/538 - Train Accuracy: 0.7375, Validation Accuracy: 0.7594, Loss: 0.3172
Epoch   7 Batch  150/538 - Train Accuracy: 0.7471, Validation Accuracy: 0.7637, Loss: 0.3146
Epoch   7 Batch  200/538 - Train Accuracy: 0.7512, Validation Accuracy: 0.7656, Loss: 0.3013
Epoch   7 Batch  250/538 - Train Accuracy: 0.7312, Validation Accuracy: 0.7658, Loss: 0.3135
Epoch   7 Batch  300/538 - Train Accuracy: 0.7467, Validation Accuracy: 0.7695, Loss: 0.2987
Epoch   7 Batch  350/538 - Train Accuracy: 0.7671, Validation Accuracy: 0.7679, Loss: 0.3018
Epoch   7 Batch  400/538 - Train Accuracy: 0.7450, Validation Accuracy: 0.7601, Loss: 0.2958
Epoch   7 Batch  450/538 - Train Accuracy: 0.7606, Validation Accuracy: 0.7727, Loss: 0.3047
Epoch   7 Batch  500/538 - Train Accuracy: 0.7688, Validation Accuracy: 0.7797, Loss: 0.2637


 40%|████      | 8/20 [07:25<11:08, 55.71s/it]

Epoch   8 Batch   50/538 - Train Accuracy: 0.7434, Validation Accuracy: 0.7800, Loss: 0.2882
Epoch   8 Batch  100/538 - Train Accuracy: 0.7482, Validation Accuracy: 0.7797, Loss: 0.2792
Epoch   8 Batch  150/538 - Train Accuracy: 0.7686, Validation Accuracy: 0.7754, Loss: 0.2759
Epoch   8 Batch  200/538 - Train Accuracy: 0.7676, Validation Accuracy: 0.7846, Loss: 0.2676
Epoch   8 Batch  250/538 - Train Accuracy: 0.7441, Validation Accuracy: 0.7743, Loss: 0.2759
Epoch   8 Batch  300/538 - Train Accuracy: 0.7602, Validation Accuracy: 0.7910, Loss: 0.2649
Epoch   8 Batch  350/538 - Train Accuracy: 0.8039, Validation Accuracy: 0.7992, Loss: 0.2704
Epoch   8 Batch  400/538 - Train Accuracy: 0.7757, Validation Accuracy: 0.8068, Loss: 0.2608
Epoch   8 Batch  450/538 - Train Accuracy: 0.7958, Validation Accuracy: 0.8001, Loss: 0.2716
Epoch   8 Batch  500/538 - Train Accuracy: 0.8059, Validation Accuracy: 0.8063, Loss: 0.2233


 45%|████▌     | 9/20 [08:21<10:12, 55.67s/it]

Epoch   9 Batch   50/538 - Train Accuracy: 0.8045, Validation Accuracy: 0.8066, Loss: 0.2446
Epoch   9 Batch  100/538 - Train Accuracy: 0.7963, Validation Accuracy: 0.8141, Loss: 0.2333
Epoch   9 Batch  150/538 - Train Accuracy: 0.8143, Validation Accuracy: 0.8169, Loss: 0.2409
Epoch   9 Batch  200/538 - Train Accuracy: 0.8381, Validation Accuracy: 0.8224, Loss: 0.2274
Epoch   9 Batch  250/538 - Train Accuracy: 0.8023, Validation Accuracy: 0.8221, Loss: 0.2338
Epoch   9 Batch  300/538 - Train Accuracy: 0.8235, Validation Accuracy: 0.8386, Loss: 0.2239
Epoch   9 Batch  350/538 - Train Accuracy: 0.8464, Validation Accuracy: 0.8324, Loss: 0.2312
Epoch   9 Batch  400/538 - Train Accuracy: 0.8296, Validation Accuracy: 0.8398, Loss: 0.2235
Epoch   9 Batch  450/538 - Train Accuracy: 0.8358, Validation Accuracy: 0.8446, Loss: 0.2309
Epoch   9 Batch  500/538 - Train Accuracy: 0.8549, Validation Accuracy: 0.8279, Loss: 0.1978


 50%|█████     | 10/20 [09:16<09:16, 55.66s/it]

Epoch  10 Batch   50/538 - Train Accuracy: 0.8404, Validation Accuracy: 0.8457, Loss: 0.2145
Epoch  10 Batch  100/538 - Train Accuracy: 0.8367, Validation Accuracy: 0.8519, Loss: 0.2113
Epoch  10 Batch  150/538 - Train Accuracy: 0.8430, Validation Accuracy: 0.8466, Loss: 0.2019
Epoch  10 Batch  200/538 - Train Accuracy: 0.8561, Validation Accuracy: 0.8400, Loss: 0.1940
Epoch  10 Batch  250/538 - Train Accuracy: 0.8457, Validation Accuracy: 0.8603, Loss: 0.1883
Epoch  10 Batch  300/538 - Train Accuracy: 0.8480, Validation Accuracy: 0.8567, Loss: 0.1826
Epoch  10 Batch  350/538 - Train Accuracy: 0.8495, Validation Accuracy: 0.8493, Loss: 0.1868
Epoch  10 Batch  400/538 - Train Accuracy: 0.8499, Validation Accuracy: 0.8569, Loss: 0.1810
Epoch  10 Batch  450/538 - Train Accuracy: 0.8477, Validation Accuracy: 0.8587, Loss: 0.1837
Epoch  10 Batch  500/538 - Train Accuracy: 0.8734, Validation Accuracy: 0.8638, Loss: 0.1540


 55%|█████▌    | 11/20 [10:12<08:21, 55.69s/it]

Epoch  11 Batch   50/538 - Train Accuracy: 0.8729, Validation Accuracy: 0.8596, Loss: 0.1685
Epoch  11 Batch  100/538 - Train Accuracy: 0.8633, Validation Accuracy: 0.8723, Loss: 0.1600
Epoch  11 Batch  150/538 - Train Accuracy: 0.8691, Validation Accuracy: 0.8706, Loss: 0.1595
Epoch  11 Batch  200/538 - Train Accuracy: 0.8908, Validation Accuracy: 0.8750, Loss: 0.1538
Epoch  11 Batch  250/538 - Train Accuracy: 0.8756, Validation Accuracy: 0.8709, Loss: 0.1656
Epoch  11 Batch  300/538 - Train Accuracy: 0.8817, Validation Accuracy: 0.8835, Loss: 0.1538
Epoch  11 Batch  350/538 - Train Accuracy: 0.8841, Validation Accuracy: 0.8786, Loss: 0.1571
Epoch  11 Batch  400/538 - Train Accuracy: 0.8942, Validation Accuracy: 0.8855, Loss: 0.1517
Epoch  11 Batch  450/538 - Train Accuracy: 0.8795, Validation Accuracy: 0.8786, Loss: 0.1606
Epoch  11 Batch  500/538 - Train Accuracy: 0.8915, Validation Accuracy: 0.8725, Loss: 0.1333


 60%|██████    | 12/20 [11:08<07:25, 55.68s/it]

Epoch  12 Batch   50/538 - Train Accuracy: 0.8729, Validation Accuracy: 0.8720, Loss: 0.1391
Epoch  12 Batch  100/538 - Train Accuracy: 0.8967, Validation Accuracy: 0.9020, Loss: 0.1366
Epoch  12 Batch  150/538 - Train Accuracy: 0.8846, Validation Accuracy: 0.8912, Loss: 0.1452
Epoch  12 Batch  200/538 - Train Accuracy: 0.8975, Validation Accuracy: 0.8887, Loss: 0.1378
Epoch  12 Batch  250/538 - Train Accuracy: 0.8867, Validation Accuracy: 0.8949, Loss: 0.1420
Epoch  12 Batch  300/538 - Train Accuracy: 0.8964, Validation Accuracy: 0.8952, Loss: 0.1410
Epoch  12 Batch  350/538 - Train Accuracy: 0.8930, Validation Accuracy: 0.8979, Loss: 0.1373
Epoch  12 Batch  400/538 - Train Accuracy: 0.9115, Validation Accuracy: 0.8995, Loss: 0.1311
Epoch  12 Batch  450/538 - Train Accuracy: 0.8888, Validation Accuracy: 0.8977, Loss: 0.1457
Epoch  12 Batch  500/538 - Train Accuracy: 0.9075, Validation Accuracy: 0.8940, Loss: 0.1156


 65%|██████▌   | 13/20 [12:03<06:28, 55.47s/it]

Epoch  13 Batch   50/538 - Train Accuracy: 0.8920, Validation Accuracy: 0.8942, Loss: 0.1258
Epoch  13 Batch  100/538 - Train Accuracy: 0.9115, Validation Accuracy: 0.9006, Loss: 0.1246
Epoch  13 Batch  150/538 - Train Accuracy: 0.9049, Validation Accuracy: 0.9034, Loss: 0.1245
Epoch  13 Batch  200/538 - Train Accuracy: 0.9104, Validation Accuracy: 0.9102, Loss: 0.1128
Epoch  13 Batch  250/538 - Train Accuracy: 0.9107, Validation Accuracy: 0.9052, Loss: 0.1222
Epoch  13 Batch  300/538 - Train Accuracy: 0.9066, Validation Accuracy: 0.9064, Loss: 0.1174
Epoch  13 Batch  350/538 - Train Accuracy: 0.9068, Validation Accuracy: 0.9071, Loss: 0.1157
Epoch  13 Batch  400/538 - Train Accuracy: 0.9163, Validation Accuracy: 0.9089, Loss: 0.1172
Epoch  13 Batch  450/538 - Train Accuracy: 0.9023, Validation Accuracy: 0.9116, Loss: 0.1224
Epoch  13 Batch  500/538 - Train Accuracy: 0.9164, Validation Accuracy: 0.9080, Loss: 0.0953


 70%|███████   | 14/20 [12:57<05:31, 55.31s/it]

Epoch  14 Batch   50/538 - Train Accuracy: 0.8998, Validation Accuracy: 0.9057, Loss: 0.1076
Epoch  14 Batch  100/538 - Train Accuracy: 0.9219, Validation Accuracy: 0.9183, Loss: 0.1041
Epoch  14 Batch  150/538 - Train Accuracy: 0.9090, Validation Accuracy: 0.9077, Loss: 0.1064
Epoch  14 Batch  200/538 - Train Accuracy: 0.9236, Validation Accuracy: 0.9064, Loss: 0.0991
Epoch  14 Batch  250/538 - Train Accuracy: 0.9270, Validation Accuracy: 0.9052, Loss: 0.1116
Epoch  14 Batch  300/538 - Train Accuracy: 0.9081, Validation Accuracy: 0.9135, Loss: 0.1021
Epoch  14 Batch  350/538 - Train Accuracy: 0.9118, Validation Accuracy: 0.9025, Loss: 0.1034
Epoch  14 Batch  400/538 - Train Accuracy: 0.9228, Validation Accuracy: 0.9086, Loss: 0.1067
Epoch  14 Batch  450/538 - Train Accuracy: 0.9076, Validation Accuracy: 0.9061, Loss: 0.1157
Epoch  14 Batch  500/538 - Train Accuracy: 0.9222, Validation Accuracy: 0.9077, Loss: 0.0886


 75%|███████▌  | 15/20 [13:52<04:35, 55.17s/it]

Epoch  15 Batch   50/538 - Train Accuracy: 0.9148, Validation Accuracy: 0.9222, Loss: 0.0952
Epoch  15 Batch  100/538 - Train Accuracy: 0.9318, Validation Accuracy: 0.9201, Loss: 0.0949
Epoch  15 Batch  150/538 - Train Accuracy: 0.9184, Validation Accuracy: 0.9206, Loss: 0.0941
Epoch  15 Batch  200/538 - Train Accuracy: 0.9264, Validation Accuracy: 0.9141, Loss: 0.0876
Epoch  15 Batch  250/538 - Train Accuracy: 0.9357, Validation Accuracy: 0.9125, Loss: 0.0969
Epoch  15 Batch  300/538 - Train Accuracy: 0.9254, Validation Accuracy: 0.9190, Loss: 0.0902
Epoch  15 Batch  350/538 - Train Accuracy: 0.9340, Validation Accuracy: 0.9226, Loss: 0.0947
Epoch  15 Batch  400/538 - Train Accuracy: 0.9379, Validation Accuracy: 0.9160, Loss: 0.0915
Epoch  15 Batch  450/538 - Train Accuracy: 0.9174, Validation Accuracy: 0.9260, Loss: 0.1037
Epoch  15 Batch  500/538 - Train Accuracy: 0.9387, Validation Accuracy: 0.9199, Loss: 0.0700


 80%|████████  | 16/20 [14:47<03:40, 55.17s/it]

Epoch  16 Batch   50/538 - Train Accuracy: 0.9217, Validation Accuracy: 0.9313, Loss: 0.0766
Epoch  16 Batch  100/538 - Train Accuracy: 0.9410, Validation Accuracy: 0.9242, Loss: 0.0711
Epoch  16 Batch  150/538 - Train Accuracy: 0.9137, Validation Accuracy: 0.9125, Loss: 0.0821
Epoch  16 Batch  200/538 - Train Accuracy: 0.9344, Validation Accuracy: 0.9300, Loss: 0.0716
Epoch  16 Batch  250/538 - Train Accuracy: 0.9338, Validation Accuracy: 0.9226, Loss: 0.0747
Epoch  16 Batch  300/538 - Train Accuracy: 0.9343, Validation Accuracy: 0.9256, Loss: 0.0738
Epoch  16 Batch  350/538 - Train Accuracy: 0.9284, Validation Accuracy: 0.9252, Loss: 0.0800
Epoch  16 Batch  400/538 - Train Accuracy: 0.9423, Validation Accuracy: 0.9277, Loss: 0.0753
Epoch  16 Batch  450/538 - Train Accuracy: 0.9142, Validation Accuracy: 0.9265, Loss: 0.0868
Epoch  16 Batch  500/538 - Train Accuracy: 0.9466, Validation Accuracy: 0.9221, Loss: 0.0626


 85%|████████▌ | 17/20 [15:42<02:45, 55.12s/it]

Epoch  17 Batch   50/538 - Train Accuracy: 0.9305, Validation Accuracy: 0.9292, Loss: 0.0669
Epoch  17 Batch  100/538 - Train Accuracy: 0.9539, Validation Accuracy: 0.9263, Loss: 0.0659
Epoch  17 Batch  150/538 - Train Accuracy: 0.9404, Validation Accuracy: 0.9373, Loss: 0.0648
Epoch  17 Batch  200/538 - Train Accuracy: 0.9402, Validation Accuracy: 0.9327, Loss: 0.0622
Epoch  17 Batch  250/538 - Train Accuracy: 0.9500, Validation Accuracy: 0.9403, Loss: 0.0666
Epoch  17 Batch  300/538 - Train Accuracy: 0.9423, Validation Accuracy: 0.9261, Loss: 0.0685
Epoch  17 Batch  350/538 - Train Accuracy: 0.9308, Validation Accuracy: 0.9217, Loss: 0.0703
Epoch  17 Batch  400/538 - Train Accuracy: 0.9436, Validation Accuracy: 0.9370, Loss: 0.0632
Epoch  17 Batch  450/538 - Train Accuracy: 0.9174, Validation Accuracy: 0.9343, Loss: 0.0736
Epoch  17 Batch  500/538 - Train Accuracy: 0.9549, Validation Accuracy: 0.9434, Loss: 0.0503


 90%|█████████ | 18/20 [16:37<01:50, 55.06s/it]

Epoch  18 Batch   50/538 - Train Accuracy: 0.9250, Validation Accuracy: 0.9387, Loss: 0.0632
Epoch  18 Batch  100/538 - Train Accuracy: 0.9463, Validation Accuracy: 0.9382, Loss: 0.0515
Epoch  18 Batch  150/538 - Train Accuracy: 0.9484, Validation Accuracy: 0.9364, Loss: 0.0600
Epoch  18 Batch  200/538 - Train Accuracy: 0.9430, Validation Accuracy: 0.9370, Loss: 0.0497
Epoch  18 Batch  250/538 - Train Accuracy: 0.9543, Validation Accuracy: 0.9318, Loss: 0.0606
Epoch  18 Batch  300/538 - Train Accuracy: 0.9382, Validation Accuracy: 0.9455, Loss: 0.0556
Epoch  18 Batch  350/538 - Train Accuracy: 0.9507, Validation Accuracy: 0.9384, Loss: 0.0638
Epoch  18 Batch  400/538 - Train Accuracy: 0.9418, Validation Accuracy: 0.9405, Loss: 0.0575
Epoch  18 Batch  450/538 - Train Accuracy: 0.9228, Validation Accuracy: 0.9318, Loss: 0.0674
Epoch  18 Batch  500/538 - Train Accuracy: 0.9602, Validation Accuracy: 0.9366, Loss: 0.0445


 95%|█████████▌| 19/20 [17:32<00:55, 55.01s/it]

Epoch  19 Batch   50/538 - Train Accuracy: 0.9436, Validation Accuracy: 0.9395, Loss: 0.0508
Epoch  19 Batch  100/538 - Train Accuracy: 0.9582, Validation Accuracy: 0.9498, Loss: 0.0474
Epoch  19 Batch  150/538 - Train Accuracy: 0.9490, Validation Accuracy: 0.9432, Loss: 0.0497
Epoch  19 Batch  200/538 - Train Accuracy: 0.9482, Validation Accuracy: 0.9428, Loss: 0.0442
Epoch  19 Batch  250/538 - Train Accuracy: 0.9563, Validation Accuracy: 0.9368, Loss: 0.0518
Epoch  19 Batch  300/538 - Train Accuracy: 0.9522, Validation Accuracy: 0.9464, Loss: 0.0521
Epoch  19 Batch  350/538 - Train Accuracy: 0.9501, Validation Accuracy: 0.9426, Loss: 0.0596
Epoch  19 Batch  400/538 - Train Accuracy: 0.9420, Validation Accuracy: 0.9407, Loss: 0.0536
Epoch  19 Batch  450/538 - Train Accuracy: 0.9252, Validation Accuracy: 0.9439, Loss: 0.0621
Epoch  19 Batch  500/538 - Train Accuracy: 0.9640, Validation Accuracy: 0.9350, Loss: 0.0368


100%|██████████| 20/20 [18:27<00:00, 54.91s/it]


Model Trained and Saved


### Hyperparameter selections and results

|embed_input|embed_output|batch_size| rnn_size|keep_prob|    lr|epochs|training_time| train_acc| val_acc|  loss|
|--------------------------------------------------------------------------------------------------------------|
|         30|          30|       128|       50|     0.50|0.0010|    10|          ?  |    0.6684|    0.7248|0.3571|
|         15|          15|       256|      100|     0.50|0.0010|    10|        09:10|    0.7305|    0.7228|0.3687|
|         15|          15|       512|      100|     0.75|0.0010|    10|        04:56|    0.6604|    0.6671|0.4456|
|         15|          15|       512|      100|     0.75|0.0005|    10|        04:53|    0.6127|    0.6345|0.6592|
|         15|          15|       512|      100|     0.75|0.0005|    30|        14:50|    0.8104|    0.7900|0.2596|
|         15|          15|       256|      100|     0.75|0.0010|    20|        18:02|    0.8766|    0.8699|0.1053|
|         15|          15|       256|      150|     0.75|0.0010|    20|        18:27|__0.9640__|__0.9350__|0.0368|

### Analysis

It appears that the biggest gains in terms of accuracy came from increasing the __rnn_size__, keep_prob, and epochs.
The __rnn_size__ seems especially important.  Increasing the batch size seems to speed training, but it may cost accuracy.

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [26]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [28]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    sentence = sentence.lower()
    seq = [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split()]
    
    return seq


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [29]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [71, 156, 125, 129, 178, 75, 43]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [357, 228, 105, 331, 331, 198, 1]
  French Words: est a vu aux aux . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.